File: leetcode/654.maximum-binary-tree/cpp/constructmaximumbinarytree.cpp
#include <vector>
#include <algorithm>

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    TreeNode *constructMaximumBinaryTree(std::vector<int> &nums)
    {
        auto prev = TreeNode();
        traversal(nums, &prev, true, 0, nums.size());
        return prev.left;
    }

private:
    void traversal(std::vector<int> &nums, TreeNode *prev, bool isLeft, int left, int right)
    {
        auto maxIdx = findMaxIdx(nums, left, right);
        if (maxIdx == -1)
            return;

        auto node = new TreeNode(nums[maxIdx]);
        if (isLeft)
            prev->left = node;
        else
            prev->right = node;

        traversal(nums, node, true, left, maxIdx);
        traversal(nums, node, false, maxIdx + 1, right);
    }

    int findMaxIdx(std::vector<int> &nums, int left, int right)
    {
        if (left >= right)
            return -1;

        auto it = std::max_element(nums.begin() + left, nums.begin() + right);
        return it - nums.begin();
    }
};

File: leetcode/654.maximum-binary-tree/ts/constructmaximumbinarytree.ts
class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}

const constructMaximumBinaryTree = (nums: number[]): TreeNode | null => {
    const prev = new TreeNode()
    traversal(nums, prev, false, 0, nums.length)
    return prev.right
}

const traversal = (nums: number[], prev: TreeNode, isLeft: boolean, left: number, right: number) => {
    const maxIdx = findMaxIdx(nums, left, right)
    if (maxIdx === null) return

    const node = new TreeNode(nums[maxIdx])

    if (isLeft) prev.left = node
    else prev.right = node

    traversal(nums, node, true, left, maxIdx)
    traversal(nums, node, false, maxIdx + 1, right)
}

const findMaxIdx = (nums: number[], left: number, right: number): number | null => {
    if (left >= right) return null

    let maxIdx = left
    for (let i = left; i < right; i++)
        if (nums[i] > nums[maxIdx]) maxIdx = i

    return maxIdx
}

export default constructMaximumBinaryTree
